crates/sui-indexer-alt-graphql/src/api/types/unchanged_shared_object.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use async_graphql::{Context, Enum, Object, SimpleObject, Union};
+use sui_types::{
+    base_types::{ObjectID, SequenceNumber},
+    digests::ObjectDigest,
+    effects::UnchangedSharedKind as NativeUnchangedSharedKind,
+};
+
+use crate::{
+    api::scalars::{sui_address::SuiAddress, uint53::UInt53},
+    error::RpcError,
+    scope::Scope,
+};
+
+use super::{
+    address::Address,
+    object::{self, Object},
+};
+
+/// Reason why a transaction was cancelled.
+#[derive(Enum, Copy, Clone, Eq, PartialEq)]
+pub(crate) enum CancellationReason {
+    /// Read operation was cancelled.
+    CancelledRead,
+    /// Object congestion prevented execution.
+    Congested,
+    /// Randomness service was unavailable.
+    RandomnessUnavailable,
+    /// Internal use only.
+    Unknown,
+}
+
+/// Details pertaining to consensus-managed objects that are referenced by but not changed by a transaction.
+#[derive(Union)]
+pub(crate) enum UnchangedSharedObject {
+    Read(ConsensusRead),
+    MutateStreamEnded(ConsensusMutateStreamEnded),
+    ReadStreamEnded(ConsensusReadStreamEnded),
+    Cancelled(ConsensusCancelled),
+    PerEpochConfig(PerEpochConfig),
+}
+
+/// A consensus-managed object that was read by this transaction but not modified.
+pub(crate) struct ConsensusRead {
+    scope: Scope,
+    object_id: ObjectID,
+    version: SequenceNumber,
+    digest: ObjectDigest,
+}
+
+#[Object]
+impl ConsensusRead {
+    /// The version of the consensus-managed object that was read by this transaction.
+    async fn object(&self) -> Option<Object> {
+        let address = Address::with_address(self.scope.clone(), self.object_id.into());
+        Some(Object::with_ref(address, self.version, self.digest))
+    }
+}
+
+/// A transaction that wanted to mutate a consensus-managed object but couldn't because it became not-consensus-managed before the transaction executed (either it was deleted, or turned into an owned object).
+#[derive(SimpleObject)]
+pub(crate) struct ConsensusMutateStreamEnded {
```

Comment:
These names are a bit mixed up imo.

There is not such a thing as a "mutate stream" or "read stream", rather it's a "consensus stream" that ended when you were trying to mutate/read the object. So the words "ConsensusStream" should stay together in the name. I would just match the native name as closely as possibile with these to make the source obvious